src/components/TimeEntries.tsx

```diff
@@ -0,0 +1,161 @@
+    hour: "2-digit",
+    minute: "2-digit",
+  });
+  const endTime = endDateTime.toLocaleTimeString("nl-NL", {
+    hour: "2-digit",
+    minute: "2-digit",
+  });
+  return { timeRegistration: `${startTime} - ${endTime}`, startDateTime, endDateTime };
+};
+
+const calculateDuration = (timeEntry: Type.TimeEntry) => {
+  const startDateTime = new Date(timeEntry.startTime);
+  const endDateTime = new Date(timeEntry.endTime);
+  const durationAsNumber = endDateTime.getTime() - startDateTime.getTime();
+  return { durationAsNumber, duration: new Date(durationAsNumber).toISOString().substr(11, 5) };
+};
+
+function TimeEntries() {
+  const rawTimeEntries = useStore((state) => state.rawTimeEntries);
+  const setRawTimeEntries = useStore((state) => state.setRawTimeEntries);
+  const [timeEntries, setTimeEntries] = useState<Type.TimeEntry[]>([]);
+
+  async function deleteTimeEntry(_id: number) {
+    const newTimeEntries = rawTimeEntries.filter((entry) => entry.id !== _id);
+    setRawTimeEntries(newTimeEntries);
+    await deleteTimeEntries(_id);
+  }
+
+  useEffect(() => {
+    const newTimeEntries = [];
+    rawTimeEntries.map((timeEntry) =>
+      newTimeEntries.push({
+        id: timeEntry.id,
+        client: timeEntry.client,
+        activity: timeEntry.activity,
+        timeRegistration: convertDates(timeEntry).timeRegistration,
+        startDateTime: convertDates(timeEntry).startDateTime,
+        endDateTime: convertDates(timeEntry).startDateTime,
+        duration: calculateDuration(timeEntry).duration,
+        durationAsNumber: calculateDuration(timeEntry).durationAsNumber,
+      }),
+    );
+    setTimeEntries(newTimeEntries);
+  }, [rawTimeEntries]);
+
+  const dateDisplay: {
+    day?: "numeric" | "2-digit";
+    month?: "numeric" | "2-digit" | "long" | "short" | "narrow";
+    year?: "numeric" | "2-digit";
+  } = {
+    day: "numeric",
+    month: "long",
+    year: "numeric",
+  };
+
+  const timeEntryFilter = useStore((state) => state.timeEntryFilter);
+
+  const fields = Object.keys(timeEntryFilter || {});
+
+  const dates = new Set<string>();
+
+  timeEntries
+    .filter((timeEntry) => fields.every((field) => timeEntryFilter[field] === timeEntry[field]))
+    .map((timeEntry) =>
+      dates.add(timeEntry.startDateTime.toLocaleDateString("nl-NL", dateDisplay)),
+    );
+
+  const dateArray = Array.from(dates);
+
+  const dateContainer = dateArray.map((date) => {
+    const filteredTimeEntries = timeEntries.filter(
+      (timeEntry) => timeEntry.startDateTime.toLocaleDateString("nl-NL", dateDisplay) === date,
+    );
+    return {
+      dateInfo: {
+        date,
+        duration: `${
+          filteredTimeEntries.reduce((acc, timeEntry) => acc + timeEntry.durationAsNumber, 0) /
+          3600000
+        } hours`,
+      },
+      timeEntryList: filteredTimeEntries,
+    };
+  });
+
+  return (
+    <Styled.TimeEntries>
+      {dateContainer.map((item) => (
```

Comment:
You should iterate over the time entries, instead of dates, this will save you many lines of code and will make everything more readable. Then you won't need to have `rawTimeEntries`, but just `timeEntries`, which can simply be passed as prop then.

To get you started:
```
{timeEntries.map((timeEntry, index) => (
  {/* Compare previous time entry, if it exists, with the current and if it's a new date: */ && (
    <DateHeading timeEntries={timeEntries.filter(timeEntry_ => /* filter to current date */)} />
  )}
  <TimeEntry key={timeEntry.id} timeEntry={timeEntry} firstChild={/* ... */} lastChild={/* ... */} /* ... */ />
))}
```

Reply:
Will be on backlog, I think I already had this before actually, but wanted to abstract it further.